modules/core/src/main/scala/ch/epfl/scala/debugadapter/internal/evaluator/RuntimeDefaultValidator.scala

```diff
@@ -304,6 +297,39 @@ class RuntimeDefaultValidator(val frame: JdiFrame, val sourceLookUp: SourceLookU
       )
     } yield ifTree
   }
+
+  /* -------------------------------------------------------------------------- */
+  /*                              Assign validation                             */
+  /* -------------------------------------------------------------------------- */
+  def isMutable(tree: RuntimeEvaluableTree): Boolean =
+    tree match {
+      case field: FieldTree => !field.immutable
+      case localVar: LocalVarTree => true
+      case _ => false
+    }
+
+  def validateAssign(
+      tree: Term.Assign,
+      localVarValidation: String => Validation[RuntimeEvaluableTree] = localVarTreeByName,
+      fieldValidation: (Validation[RuntimeTree], String) => Validation[RuntimeEvaluableTree] = fieldTreeByName
+  ): Validation[RuntimeEvaluableTree] = {
+    val lhs = tree.lhs match {
+      case select: Term.Select =>
+        fieldValidation(validateWithClass(select.qual), select.name.value)
+      case name: Term.Name =>
+        localVarValidation(name.value).orElse(fieldValidation(thisTree, name.value))
+      case _ => Recoverable("Unsupported assignment")
+    }
```

Comment:
I wonder if it would not be simpler to pass a boolean to `filedTreeByName` and `localVarTreeByName` to prevent them from pre-evaluating:

```suggestion
  def validateAssign(tree: Term.Assign): Validation[RuntimeEvaluableTree] = {
    val lhs = tree.lhs match {
      case select: Term.Select =>
        fieldTreeByName(validateWithClass(select.qual), select.name.value, preEvaluate = false)
      case name: Term.Name =>
        localVarTreeByName(name.value, preEvaluate = false).orElse(fieldTreeByName(thisTree, name.value, preEvaluate = false))
      case _ => Recoverable("Unsupported assignment")
    }
```

Reply:
Yep, seems simpler